Add value to Protect card - Cleanup, refactor, & final enhancements.

elliottprogrammer opened the pull request:
This PR refactors and cleans up the code/components that render the Protect card value section in My Jetpack.

<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->
See Design P2: p1HpG7-rFA-p2 (Figma link is within the post)

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
- Update: Auto-Firewall column: Show tooltip when user has a paid plan but the auto-firewall is manually turned off/disabled.
- Add: Logins Blocked column: Add a 'success', 'inactive', and 'off' statuses (with icon and/or tooltip when applicable) to the Logins Blocked column.
- Clean up: Move/separate the "Last scan:" text code/logic into it's own `useLastScanText()` hook.
- Clean up: Remove the need to pass in `tooltipContent` as a prop into ScanStatus & ThreatStatus components. Just call the `useProtectTooltipCopy()` hook within the components instead.
- Clean up: Remove the need to pass in props into the `useProtectTooltipCopy()` hook. They are unnecessary props, just pull the data from within the hook instead.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

Project Thread: pbNhbs-aP6-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Create a Jurassic Ninja site (with options). Do not include Jetpack, but add the Beta Tester plugin and run this PR branch Jetpack Boost (or any other free-to-start standalone plugin).
* Go to My Jetpack. Connect your site. (You don't have to connect your User, if you don't want).
* Go to My Jetpack, in the Protect product card, verify the Logins Blocked column has the shield icon and reads "Off", the same as the other status icons.
![Screen Shot on 2024-07-24 at 13-40-26](https://github.com/user-attachments/assets/20d19587-ccd8-4a6c-99b5-4edf750be7a9)
* Now SSH into your JN site (using the 'ssh command' and 'password' provided in the WP Dashboard).
* Run command: `wp option update jetpack_protect_blocked_attempts 28`
* Reload My Jetpack page. Verify that the Protect card "Logins Blocked" shows a warning icon, 28,  and a tooltip. The tooltip should say, "For Brute Force Protection, activate the Jetpack or Protect plugin and enable it in settings."
    - This is basically simulating that Jetpack or Protect was once installed & activated in the past and the Brute Force Protection module blocked some logins at some point. But now the Jetpack and/or Protect plugin is not activated at the moment (therefore Brute Force Protection is not available).
![Screen Shot on 2024-07-24 at 14-44-27](https://github.com/user-attachments/assets/6ec60cad-ebb0-440d-aa80-3b1c832bdcbf)
* Now set Logins Blocked back to 0, by running command: `wp option update jetpack_protect_blocked_attempts 0`
* Go to Jetpack Beta and run this PR branch on the Jetpack Plugin.
* Go to My Jetpack and view the Protect card. Logins Blocked should show a shield icon with green checkmark, and a tooltip that says, "Brute Force Protection is actively blocking malicious login attempts. The number of blocked login attempts will display here soon!".
    - This is basically simulating the case where Brute Force Protection is on and working, but no logins have been blocked yet; (# of logins blocked = 0).
![Screen Shot on 2024-07-24 at 13-55-29](https://github.com/user-attachments/assets/7bfea8ac-2426-461e-9ec9-8bdc9f80a361)
* Now run command: `wp option update jetpack_protect_blocked_attempts 28`
* Reload My Jetpack, view the Protect card. Logins Blocked should show "28".  (No tooltip)
![Screen Shot on 2024-07-24 at 13-57-04](https://github.com/user-attachments/assets/0d7a2885-a352-42e6-ad0f-3bda6a7f339b)
* Now go to Jetpack --> Settings, toggle Off Brute Force Protection.
* Go to My Jetpack, View the Protect card. The Logins Blocked should show a warning icon, "28",  and a tooltip that says, "Brute Force Protection is disabled and not actively blocking malicious login attempts. Go to <a>Jetpack settings</a> to activate it.".  The settings link should point to Jetpack Settings (`admin.php?page=jetpack#/settings`).
![Screen Shot on 2024-07-24 at 13-58-10](https://github.com/user-attachments/assets/fc3952ba-376f-4b62-b4d4-76002b9a927d)
* Now go to the Plugins page, and Deactivate the Jetpack Plugin. Then go to the Beta plugin and add this PR branch to the Protect plugin.
* Go to My Jetpack. The Logins Blocked should show "28".  No tooltip. (Because Brute Force Protection was automatically turned on when activating the Protect plugin. This is expected).
![Screen Shot on 2024-07-24 at 14-16-05](https://github.com/user-attachments/assets/3660d803-d234-458f-9de6-643bef3be1e6)
* Click "View" on the Protect card to go to the Protect page. Click on the "Firewall" tab. Toggle "Off" Brute Force Protection.
* Go to My Jetpack. The Logins Blocked should now show a warning icon, "28", and a tooltip again. that says, "Brute Force Protection is disabled and not actively blocking malicious login attempts. Go to <a>firewall settings</a> to activate it.".  The settings link should this time point to **Protect firewall** Settings (`admin.php?page=jetpack-protect#/firewall`).
![Screen Shot on 2024-07-24 at 14-15-14](https://github.com/user-attachments/assets/134ad584-edd9-4b83-a9bf-46622ed9b6aa)
* Set Logins Blocked back to 0 again by running command: `wp option update jetpack_protect_blocked_attempts 0`
* Reload My Jetpack page, View the Protect card: Logins Blocked should be showing a warning shield icon, with Tooltip that says the same as above again:  "Brute Force Protection is disabled and not actively blocking malicious login attempts. Go to <a>firewall settings</a> to activate it.".  The settings link should point to Protect firewall Settings (`admin.php?page=jetpack-protect#/firewall`).
![Screen Shot on 2024-07-24 at 14-19-42](https://github.com/user-attachments/assets/1362ea13-853b-40b0-b756-09d87d08295d)
* Now view the "Auto-Firewall" column:  It should currently be "inactive" with tooltip that says, "Upgrade required for activation. Manual rules available."
![Screen Shot on 2024-07-24 at 14-20-55](https://github.com/user-attachments/assets/f2f45adc-bf29-47b2-aaac-225e0512a282)
* On the Protect card, click the "Upgrade" button. Purchase the upgrade plan. (using a8c credits). (Or add /jetpack_scan (Jetpack Scan Daily) to the site in Store Admin).
* After purchasing, go to the Protect firewall settings page.  Make sure auto-firewall is disabled.
* Go back to My Jetpack.  View the Auto-Firewall column on the Protect card.  It should show 'Inactive' **with tooltip**.  Tooltip should say, "You have Auto-Firewall disabled, visit your Protect <a>firewall settings</a> to activate." and the link should point to: `admin.php?page=jetpack-protect#/firewall`.
![Screen Shot on 2024-07-24 at 14-39-20](https://github.com/user-attachments/assets/e995c14f-533f-4b92-8166-5de4af26b20a)
And that's it!

elliottprogrammer committed:
Add the logins blocked data to the "Logins Blocked" column.

elliottprogrammer committed:
changelog

elliottprogrammer committed:
Clean up & refactor everything. Create useLastScanText hook.

elliottprogrammer committed:
Add tracking to tooltip content link click.

elliottprogrammer committed:
Add the Logins Blocked off/inactive status and tooltip copy.

elliottprogrammer committed:
Merge branch 'trunk' into update/mj-protect-card-update-n-refactor

elliottprogrammer committed:
changelog

elliottprogrammer committed:
Add settings link into Logins-Blocked tooltip.

elliottprogrammer committed:
Add tooltip link styles.

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/auto-firewall-status.tsx:
```diff
@@ -70,21 +67,19 @@ function WafStatus( { status }: { status: 'success' | 'inactive' | 'off' } ) {
+					tracksEventProps={ {
+						location: 'auto-firewall',
+						status: 'inactive',
+						hasPaidPlan: false,
```
If we are no longer checking for a paid plan here, will this still be accurate?

elliottprogrammer replied:
No it won't. Thank you, good catch!

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/logins-blocked-status.tsx:
```diff
@@ -0,0 +1,130 @@
+		plugins,
+	} = getMyJetpackWindowInitialState();
+	const { blocked_logins: blockedLoginsCount, brute_force_protection: hasBruteForceProtection } =
+		wafData;
```
```suggestion
		wafData || {};
```

Might want to add this, just in case 😄

elliottprogrammer replied:
Yes, thanks! 👍

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/logins-blocked-status.tsx:
```diff
@@ -0,0 +1,130 @@
+	const { blocked_logins: blockedLoginsCount, brute_force_protection: hasBruteForceProtection } =
+		wafData;
+
+	const isJetpackPluginActive = useMemo( () => {
+		const jetpackPlugin = Object.values( plugins ).find( plugin => plugin?.Name === 'Jetpack' );
+		return jetpackPlugin && jetpackPlugin.active;
+	}, [ plugins ] );
```
I wonder if this could be good to move to the utils folder as a reusable function 🤔

elliottprogrammer replied:
Yeah, good idea.

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/logins-blocked-status.tsx:
```diff
@@ -0,0 +1,130 @@
+ */
+function BlockedStatus( {
+	status,
+	blockedLoginsCount,
```
It looks like the blockedLoginsCount is being passed to this function every time. Could we get that from the state in this function instead to avoid unnecessary props?

elliottprogrammer replied:
yes, good point, good idea. 👍

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/logins-blocked-status.tsx:
```diff
@@ -0,0 +1,130 @@
+
+	if ( isPluginActive && isSiteConnected ) {
+		if ( hasBruteForceProtection ) {
+			return <BlockedStatus status="success" blockedLoginsCount={ blockedLoginsCount } />;
```
Should this status maybe be "active" instead of "success"? I think that may be a more accurate descriptor

elliottprogrammer replied:
Yeah, good point. I'll use the status string of "active" for both Logins Blocked and Auto Firewall components, as that makes more sense for these.  Although the status string of "success" still works well for ScanStatus though, so I'll leave that one unchanged. 👍

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/logins-blocked-status.tsx:
```diff
@@ -0,0 +1,130 @@
+					tracksEventName={ 'protect_card_tooltip_open' }
+					tracksEventProps={ {
+						location: 'blocked-logins',
+						status: 'success',
```
If you do update the status, we'll want to do that here as well. Also we could probably just replace the hardcoded value here with the variable itself 😄

CodeyGuyDylan replied:
```suggestion
						status,
```

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/use-protect-tooltip-copy.ts:
```diff
@@ -13,24 +20,136 @@ export type TooltipContent = {
 	const { detail } = useProduct( slug );
-	const { hasPaidPlanForProduct: hasProtectPaidPlan } = detail;
+	const { isPluginActive: isProtectPluginActive, hasPaidPlanForProduct: hasProtectPaidPlan } =
+		detail;
```
```suggestion
		detail || {};
```

Probably have this here just to be safe 😄. Also should probably have this for the `scanData` and `wafData` below

elliottprogrammer replied:
Thanks!  Yes, I handled every occurrence of this when destructuring myJetpackInitialState, throughout all these Protect card files. 👍

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-cards-section/protect-card/use-protect-tooltip-copy.ts:
```diff
@@ -82,9 +201,30 @@ export function useProtectTooltipCopy( {
+							),
+							{
+								a: createElement( 'a', {
+									href: 'admin.php?page=jetpack-protect#/firewall',
```
Is there any reason we can't use the `settingsLink` variable here as well? If we are sending them to a protect admin page I assume that Jetpack Protect is installed, and if that is the case, settingsLink should be the same value as we have hardcoded here

elliottprogrammer replied:
Ahh yes, thanks.  I overlooked this one.

CodeyGuyDylan reviewed (changes requested):
I think this looks great! The testing instructions worked as described. I think a few screenshots may be helpful in following along but otherwise very well described

> 	Go to My Jetpack and view the Protect card. Logins Blocked should show "0", with tooltip that says, "Brute Force Protection is actively blocking malicious login attempts. Data will display here soon!" (Tooltip is matching what is displayed in the Jetpack Dashboard (The Brute Force Protection card)).

When the site is connected but not the user, the Dashboard does not show this text for Brute Force Protection, it shows a message to connect your user account

![image](https://github.com/user-attachments/assets/725701cb-60b6-40d8-ab25-fe3d739927d5)

I don’t know if that means it’s required or not so I don’t know if that’s a problem, but I thought I’d point it out

One thought about when the brute force protection is disabled but there are blocked logins

![image](https://github.com/user-attachments/assets/78fda672-722c-42ee-bc3f-88e091ade10a)

I think the info marker doesn’t really convey there is a problem (the problem being that it is disabled). This may or may not be intentional by the user but I think we should display this as at least a type of warning rather than an “info” so the user is more likely to click on it and see that it is inactive and not blocking malicious login attempts. What do you think about that? If I were looking at this as a user, I would assume that everything is on and working as intended as there are no visible warnings or errors

elliottprogrammer committed:
Add status icons to Blocked Logins in various cases.

elliottprogrammer committed:
Fix Tracks 'hasPaidPlan' prop.

elliottprogrammer committed:
Add empty object default on every occurance of destructuring myJetpackInitialState.

elliottprogrammer committed:
Move isJetpackPluginActive() to /utils directory.

elliottprogrammer committed:
Remove 'blockedLoginsCount' as prop of BlockedStatus component.

elliottprogrammer committed:
Use status string "active" instead of "success" for AutoFirewallStatus and LoginsBlockedStatus components.

elliottprogrammer committed:
Remove hardcoded 'status' tracks prop value, where applicable.

elliottprogrammer committed:
Use settingsLink var instead of hardcoded url string.

elliottprogrammer committed:
Merge branch 'trunk' into update/mj-protect-card-update-n-refactor

elliottprogrammer commented:
Thanks for your review @CodeyGuyDylan! 😃
I addressed all feedback and updated the testing instructions (and also with screenshots).  Would you mind taking another look?  Much appreciated! 🙌

CodeyGuyDylan reviewed (approved):
Code updates look good and the testing instructions worked as well. Great work on this! I think it looks awesome!
